posthog/temporal/batch_exports/squash_person_overrides.py

```diff
@@ -1,90 +1,97 @@
+PRIMARY KEY team_id, distinct_id
+SOURCE(CLICKHOUSE(USER '{user}' PASSWORD '{password}' TABLE 'person_distinct_id_overrides' DB '{database}'))
 LAYOUT(complex_key_hashed())
-LIFETIME(0)
+LIFETIME(MIN 0 MAX 0)
+"""
+
+RELOAD_DICTIONARY_QUERY = """
+SYSTEM RELOAD DICTIONARY {database}.{dictionary_name}
 """
 
 SQUASH_EVENTS_QUERY = """
 ALTER TABLE
     {database}.sharded_events
+ON CLUSTER
+    {cluster}
 UPDATE
-    person_id = dictGet('{database}.{dictionary_name}', 'override_person_id', (toInt32(team_id), person_id))
+    person_id = dictGet('{database}.{dictionary_name}', 'person_id', (team_id, distinct_id))
 IN PARTITION
     %(partition_id)s
 WHERE
-    dictHas('{database}.{dictionary_name}', (toInt32(team_id), person_id))
-    {team_id_filter}
-    AND created_at <= %(latest_created_at)s;
+    dictHas('{database}.{dictionary_name}', (team_id, distinct_id))
+    {in_team_ids}
 """
 
-DROP_DICTIONARY_QUERY = """
-DROP DICTIONARY {database}.{dictionary_name};
+SQUASH_MUTATIONS_IN_PROGRESS_QUERY = """
+SELECT mutation_id, is_done
+FROM clusterAllReplicas('{cluster}', 'system', mutations)
+WHERE table = 'sharded_events'
+AND database = '{database}'
+AND command LIKE
+    'UPDATE person_id = dictGet(''{database}.{dictionary_name}'', ''person_id'', (team_id, distinct_id)) IN PARTITION ''{partition_id}''%'
 """
 
-DELETE_SQUASHED_PERSON_OVERRIDES_QUERY = """
-ALTER TABLE
-    {database}.person_overrides
-DELETE WHERE
-    old_person_id IN %(old_person_ids)s
-    AND created_at <= %(latest_created_at)s;
+KILL_SQUASH_MUTATION_IN_PROGRESS_QUERY = """
+KILL MUTATION ON CLUSTER {cluster}
+WHERE is_done = 0
+AND table = 'sharded_events'
+AND database = '{database}'
+AND command LIKE
+    'UPDATE person_id = dictGet(''{database}.{dictionary_name}'', ''person_id'', (team_id, distinct_id)) IN PARTITION '''{partition_id}''%'
+"""
+
+DROP_DICTIONARY_QUERY = """
+DROP DICTIONARY {database}.{dictionary_name};
 """
 
-SELECT_CREATED_AT_FOR_PERSON_EVENT_QUERY = """
+CREATE_JOIN_TABLE_FOR_DELETES_QUERY = """
+CREATE TABLE {database}.person_overrides_to_delete
+ENGINE = Join(ANY, LEFT, team_id, distinct_id) AS
 SELECT
-    min(created_at) AS oldest_event_at
+    team_id, distinct_id, groupUniqArray(_partition_id) AS partitions
 FROM
     {database}.sharded_events
 WHERE
-    team_id = %(team_id)s
-    AND person_id = %(old_person_id)s
-    -- Not necessary, but can speed up query.
-    AND created_at <= %(oldest_event_at)s
-    AND created_at >= 0;
+    dictHas('{database}.{dictionary_name}', (team_id, distinct_id))
+GROUP BY
+    team_id, distinct_id
+"""
+
+DROP_JOIN_TABLE_FOR_DELETES_QUERY = """
+DROP TABLE IF EXISTS {database}.person_overrides_to_delete
+"""
+
+DELETE_SQUASHED_PERSON_OVERRIDES_QUERY = """
+ALTER TABLE
+    {database}.person_distinct_id_overrides
+DELETE WHERE
+    hasAll(joinGet('{database}.person_overrides_to_delete', 'partitions', team_id, distinct_id), %(partition_ids)s)
+    AND NOW() - _timestamp > %(grace_period)s
```

Comment:
~~Now that the overrides table isn't being detached, doesn't that mean that this statement could delete overrides that arrived _after_ the dictionary was populated if those new overrides share the same `team_id` and `distinct_id` as a value in the dictionary (and transitively the join table?)~~ I guess the grace period protects against this as long as these jobs don't take an extremely long time.

Reply:
Yup that was my thought too. And we can enforce these jobs not taking longer than the grace period via a timeout in temporal, so it should be safe.